### ISSUE TITLE
KKPT型評価関数の実装

### ISSUE THREAD

**Issue by burokoron** [burokoron/Yolts#12]
# 目的

KKPT型評価関数を実装し、この枠組みでP型評価関数の学習を行うことで、正しく実装できていることを確認する。

# 方法

[爆裂転生太郎V2.0.2](https://github.com/burokoron/Yolts/issues/9#issuecomment-1214405844)の学習に使用した棋譜を用いてKKPT型評価関数の枠組みでP型評価関数の学習を行う。KKPT評価関数のうちKとKとTを固定値にすることでP型評価関数を模倣する。また、爆裂転生太郎V2.0.2では学習にscikit-learnを用いていたが、GPUを利用できず、学習に時間がかかるため、TensorFlowを用いて学習を行う。

学習条件は以下の通りである。

- 学習方法
  - モデルアーキテクチャ
    ```mermaid
    flowchart TD
      Inputs\nn,95\n盤面81升+持ち駒7升*2=95次元 --> Embedding\nn,95,1\n入力次元数=K81升*K81升*P81升*30種*手番2種=38644290,重み初期値=0,インデックス値=0はマスク値 --> sum\nn,1 --> Outputs\nn,1\n評価値
    ```
  - 最適化関数：Adam(learning rate=0.001)
  - 学習スケジューラー
    - 1エポック以上lossが改善しない場合、学習率を0.2倍にする
    - 3エポック以上lossが改善しない場合、学習を終了する
  - バッチサイズ：4096
  - 損失関数：MSE
  - データシャッフル：なし
  - データ拡張
    - 盤面180度反転
  - 勝ち(負け)読み切り局面における評価値：6842
  - 学習時の評価値スケーリングとして評価値を512で割る
 
# 結果

[爆裂転生太郎V2.0.2](https://github.com/burokoron/Yolts/issues/9#issuecomment-1214405844)の学習で使用した3600局の棋譜で学習を行い、爆裂転生太郎V2.0.2と探索深さ4での対局勝率を比較した。また、局面の重複は除外した。結果は以下の通りである。

| 学習棋譜数 | 学習局面数(データ拡張含む) | 検証棋譜数 | 検証局面数(データ拡張含む) | MSE(スケーリング前換算MSE) |
| ---: | ---: | ---: | ---: | ---: |
| 3600 | 413006 | 400 | 59310 | 1.18(309330) |

| 仮想敵 | 勝率 |
| :--: | ---: |
| 爆裂転生太郎V2.0.2 | 46.3 |

# 結論

実験の結果、KKPT型評価関数の枠組みでP型評価関数を模倣したエンジンは爆裂転生太郎V2.0.2に対して勝率46.3%だった。勝率45％以上なので有意水準99％で有意に弱いわけではないと判断できる。したがって問題なくKKPT型評価関数の実装および学習ができることを確認できた。

## 実験時メモ

学習時のパラメータとして以下に変更するとレーティング100点ほど弱くなった。

- embedding層のインデックス0をマスク値としない
- データシャッフルをON

評価関数のパラメータを確認したところ、駒の平均価値が先手後手で非対称になっており、玉の平均価値も約9点になっていた。embedding層のインデックス0をマスク値とするように変更することで駒の平均価値が先手後手で対称となり、レーティングも50点ほど弱いようになった。さらにデータシャッフルをOFFにすることで玉の平均価値も約0点となり、レーティングに有意差はなくなった。
